platform/solr/solr-factory/src/main/java/org/codice/solr/factory/impl/SolrClientAdapter.java

```diff
@@ -0,0 +1,998 @@
+    try {
+      return code.invoke(getProxiedClient());
+    } catch (UnavailableSolrException e) {
+      throw e;
+    } catch (SolrException e) {
+      LOGGER.debug(
+          "Solr({}): API failure with code [{}] and metadata [{}]; {}",
+          core,
+          e.code(),
+          e.getMetadata(),
+          e, // this will get the info logged on the first line
+          e); // this one will get the stack trace logged after
+      checkIfReachableAndChangeStateAccordingly(e, "from the API after an error was detected");
+      throw e;
+    } catch (SolrServerException | IOException e) {
+      LOGGER.debug("Solr({}): API failure; {}", core, e, e);
+      checkIfReachableAndChangeStateAccordingly(e, "from the API after an error was detected");
+      throw e;
+    }
+  }
+
+  @Override
+  public final SolrClient getClient() {
+    // returning this to make sure all calls from our SolrClient API will still be intercepted
+    // when the client is retrieved and passed to a Solr request object
+    // this should be temporary until we completely abstract out Solr with interfaces
+    return this;
+  }
+
+  @Override
+  public String getCore() {
+    return core;
+  }
+
+  @Override
+  @SuppressWarnings("squid:S2093" /* closing of real client handled by finalizeStateChange() */)
+  public void close() throws IOException {
+    try {
+      if (state == State.CLOSED) {
+        return;
+      }
+      final SolrClient previousClientToClose;
+      final Future<?> futureToCancel;
+
+      synchronized (lock) {
+        if (state == State.CLOSED) { // already closed so bail
+          return;
+        }
+        futureToCancel = future;
+        previousClientToClose = realClient;
+        LOGGER.debug("Solr({}): closing", core);
+        this.unavailableClient =
+            new UnavailableSolrClient(
+                new UnavailableSolrException("'" + core + "' client was closed"));
+        this.apiClient = unavailableClient;
+        this.pingClient = unavailableClient;
+        this.realClient = null;
+        this.state = State.CLOSED;
+        this.future = null;
+        lock.notifyAll(); // wakeup those waiting for isAvailable(timeout)
+      }
+      finalizeStateChange(true, futureToCancel, previousClientToClose, false);
+    } finally {
+      listeners.clear();
+      initializers.clear();
+    }
+  }
+
+  @Override
+  public boolean isAvailable() {
+    final boolean available = (state == State.CONNECTED);
+
+    if (LOGGER.isDebugEnabled()) {
+      LOGGER.debug(
+          "Solr({}): current availability is [{} = {}]",
+          core,
+          state,
+          SolrClientAdapter.availableToString(available));
+    }
+    if (available && wasNotRecent(lastPing, SolrClientAdapter.PING_MIN_FREQUENCY)) {
+      LOGGER.debug(
+          "Solr({}): Proxy is starting a background task to ping the client because the last ping was too long ago",
+          core);
+      executor.submit(this::backgroundPing);
```

Comment:
That is the difference between calling `ping()` or `isAvailable()`. The later simply returns you the last known state while possibly triggering a background one if it has been too long since there was a check whereas the former does an active ping.